test/application_system_test_case.rb

```diff
@@ -18,4 +18,52 @@ def sign_in!(user = nil)
     # @current_user.confirm
     # sign_in @current_user
   end
+
+  # This adds the within option to assert_text
+  # For example:
+  # assert_text "Ruby", within: "h3.title"
+  def assert_text(text, *args, **options)
+    context = options.delete(:within)
+    if context
+      within(context) { assert_text(text, *args, **options) }
+    else
+      super
+    end
+  end
+
+  # This does a string comparison between some given HTML
+  # and some HTML found in the document.
+  def assert_html(html, within: "body")
+    # These save options give us an HTML set that's compatible
+    # between capybara and writing HTML.
+    nokogiri_save_options = Nokogiri::XML::Node::SaveOptions::FORMAT |
+                            Nokogiri::XML::Node::SaveOptions::NO_DECLARATION |
+                            Nokogiri::XML::Node::SaveOptions::AS_XHTML
+
+    # Get the html of either the body or the css path
+    context = find(:css, within)['innerHTML']
+
+    # Format the passed HTML so that it is in the same format
+    # as the HTML retrieved by the line above. To do this we
+    # remove any newlines, whitespace between tags,
+    # and any leading or trailing whitespace.
+    html.gsub!(/\n\s*/, " ")
+    html.gsub!(/>\s+</, "><")
+    html.gsub!(/^\s+/, "")
+    html.gsub!(/\s+$/, "")
+
+    # Format them both in the same manner and indent to zero to remove whitespace
+    formatted_html = Nokogiri::XML(html).to_xhtml(indent: 0, save_with: nokogiri_save_options)
+    formatted_context = Nokogiri::XML(context).to_xhtml(indent: 0, save_with: nokogiri_save_options)
+
+    # Create a useful and readable error message in case of no match
+    error_msg = lambda {
+      pretty_html = Nokogiri::XML(html).to_xhtml(indent: 2, save_with: nokogiri_save_options)
+      pretty_context = Nokogiri::XML(context).to_xhtml(indent: 2, save_with: nokogiri_save_options)
+      "Could not find:\n#{pretty_html}\n...in...#{pretty_context}"
+    }
+
+    # Run the assertion
+    assert_includes formatted_context, formatted_html, error_msg
+  end
```

Comment:
This is a complex method, but it works really well! Happy to move these into mixins etc too, but let's do that as a separate job in a separate PR 🙂

Reply:
Is this different from `assert_dom_equal`? I'm unsure though if that method works on system tests. Maybe we could import the module?

Reply:
I think the `assert_dom_equal` expects the two things to be identical, whereas this expects one to include the other. So you say "this block of HTML should appear somewhere" without having to worry about the wider structure of the page around it.